presto-raptorx/src/main/java/com/facebook/presto/raptorx/transaction/TransactionWriter.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptorx.transaction;
+
+import com.facebook.presto.raptorx.metadata.MetadataWriter;
+import org.jdbi.v3.core.JdbiException;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.OptionalLong;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static com.facebook.presto.raptorx.util.DatabaseUtil.metadataError;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class TransactionWriter
+{
+    private final MetadataWriter writer;
+    private final ReentrantLock lock = new ReentrantLock(true);
+
+    @GuardedBy("lock")
+    private boolean needRecovery = true;
+
+    @Inject
+    public TransactionWriter(MetadataWriter writer)
+    {
+        this.writer = requireNonNull(writer, "writer is null");
+    }
+
+    public void write(List<Action> actions, OptionalLong transactionId)
+    {
+        if (!actions.isEmpty()) {
+            write(() -> doWrite(actions, transactionId));
+        }
+    }
+
+    public void recover()
+    {
+        write(() -> {});
+    }
+
+    private void write(Runnable runnable)
+    {
+        lock.lock();
+        try {
+            if (needRecovery) {
+                writer.recover();
```

Comment:
"highker on Jan 30  Member
Why we recover from the next write rather than recovery directly from failure?"

So far it's logically the same. The commit is sequential, one after another. We had a doc "CodeReview" covering this. Besides, Presto's rollback logic is a lot of concern for now, cannot move the logic.